### ISSUE TITLE
[v3] improve peerDependency

### ISSUE THREAD

**Issue by kopax** [marmelab/react-admin#1683]
singletons should be a peer; `react-redux` thus probably should be

singletons (something that shares mutable state, and needs to have only one copy in the dependency graph)

Also `redux` is a way to store state. it is *definitely* a singleton.

If your lib has `react-redux` v3.0 as a dep, and someone else has v3.1, you'll have two copies in memory and things will break.

`react-redux` should be a peer dep, so should `react-router`

Also, it should be using ^ not ~ throughout, because ~ means "just patch" and ^ means "patch or minor"

You want semver ranges to be as wide as possible to allow for as much deduping and upgrading as possible.

**Comment by fzaninotto** [comment]
If someone installs another version of Redux, it’s they responsibility. If they want to share the same store, they must use the same as admin-on-rest. 

As for using ^ instead of ~, I disagree. Using ^ guarantees that the application will break easily, because lib authors don’t use delver properly. Believe me, we’ve followed this path before.

**Comment by ljharb** [comment]
@fzaninotto fwiw you're highly wrong about `^` - lib authors who don't use semver properly could publish a breaking change in a patch, too, which ~ allows. The *only* way to protect yourself from this kind of thing is with a lockfile, which is an app-level concern - published packages should use as wide a range as possible. Using `~`, or even using `=`, will also not protect against transitive dependencies from updating.

**Comment by kopax** [comment]
@fzaninotto I agree with ljharb. About peerDeps, I can only ask now why do you want such restriction ?

The customApp example let us think the opposite.

**Comment by fzaninotto** [comment]
Peer dependencies make the install process harder and more error prone. We want the setup to be easy for newcomers - sometimes at the expense of expert users like the ones who want to integrate admin-on-rest in another app.

**Comment by ljharb** [comment]
Not if you make it both a peer and a production dep - users don’t ave to do anything manually, but are prevented from having conflicts in their dep graph.

We can keep debating as long as you like, but you’re wrong here on every metric.

**Comment by kopax** [comment]
That seems like a good and hybrid solution. Can we reopen this?

**Comment by fzaninotto** [comment]
@ljharb I'm often wrong on many metrics, but it' the first time I happen to be wrong on every metric. Thanks for pointing out. By the way, which metric are you referring to? 

But since I've got an expert in the field of dependencies, let me ask a few questions, because many things remain unclear to me in npm land.

- I cannot commit a lock file because you say it's an app-wide concern, and admin-on-rest doesn't publish an app, but a set of packages. So that means that there is absolutely no way to prevent users from installing incompatible packages based on my package.json, is that right? That's my understanding, and that's the reason why I'm using ~, to prevent breakages in user's app when a dependency updates - at the cost of composability with an exising set of packages. IMO, this is a limitation of npm, and using ~ is a tradeoff to favor stability over composability.
- how does adding a dependency both as peer and prod differ from adding it as a simple dependency? If a user adds a different version of redux in their dependencies, when they'll launch the production build, they'll end up with two conflicting versions, and therefore two copies of redux. Or am I missing something?

Oh, and could you please try to be more gentle in your comments? I was close to locking you out of the repository when I read your last contribution.

Love.

**Comment by ljharb** [comment]
Regarding your first point: using ~ doesn’t prevent breakages in a user’s app, nor does using ^, nor does using =, because that only restricts your top-level deps - not any of the transitive deps. All you achieve by using ~ is a) ensuring that the deps *you have explicitly selected* can’t accidentally ship a bug in a minor version and break you, but they can still ship a bug in a patch version and break you, and the dependencies *they* select (recursively) can ship a bug in whatever semver range they’re already using. If you used ^, the previous sentence would be the same except that minor → major and patch → minor. And then b), it prevents your consumers from maximally deduping their dependency graph, causing their bundle sizes to be larger and install times to be slower, making their app slower for startup and for their users.

Regarding your second point, when it’s both a dep + a peer dep, the dep part makes it be installed automatically without user intervention; but the peer dep restriction hoists, and the user is *prevented* from having a conflicting version in their tree (they could still have a duplicate copy of the same version, but npm tries pretty hard to avoid that, and `npm dedupe` or `rm -rf node_modules && npm install` would resolve it).

I apologize that my comment was too harsh; closing an actionable issue prior to completing the discussion is a hostile action, so i was operating within that assumed context.

**Comment by fzaninotto** [comment]
I'm well aware of the tradeoff of using ~ vs ^ (stability vs deduplication). My experience is that semver isn't used enough in the JS community, so ~ is the max flexibility we can offer with an acceptable level of stability. This may change in the future, but I've been bitten too many times in the past to reconsider it for now.

As for dep + peer dep, I understand that it's a protection against bugs due to duplicated package (e.g. duplicated React), and as such it's a good idea. Would you mind opening a PR for that?

**Comment by ljharb** [comment]
That's a very surprising experience - mine is that semver is used *exclusively*, and that any dependency that knowingly violates it is summarily ejected from a dependency tree. As I've explained, using `~` doesn't give you any guarantees without a lockfile, and using a lockfile gives you complete guarantees even if you're using "*", so I'm still not sure what the concern is.

I'll let @kopax open the PR, thanks for being willing to accept it.

**Comment by fzaninotto** [comment]
> semver is used exclusively, and that any dependency that knowingly violates it is summarily ejected from a dependency tree

You want an example? material-ui, for one, didn't follow semver *at all* for the past 6 months, and broke BC on bugfix releases every week. Ejecting it is not an option.

**Comment by ljharb** [comment]
That sounds like a great reason to use `=` - not `~` - with material-ui, but I don't see how that extends to the rest of npm. The list of popular libs that violate semver is very short; I'll take note that material-ui is one of them.

**Comment by kopax** [comment]
As @ljharb suggested, I can proceed and fix peerDeps when this is re-opened. Thank you for reconsidering it.

**Comment by kopax** [comment]
Hi, sorry I didn't notice that you reopened. I will schedule this asap. Thanks for reconsidering this.

**Comment by kopax** [comment]
Any update on this? We'd like to use react-redux v6.0.0 to work a bit with the context:

https://react-redux.js.org/using-react-redux/accessing-store

This was introduced in react-redux v6.0.0

**Comment by fzaninotto** [comment]
The problem goes beyond `redux`: Every dependency we have that uses the new React context API must be a singleton (cf https://github.com/facebook/react/issues/13346). And that's a lot: `react-redux`, `react-router`, `material-ui`, `redux-form`, `react-transition-group`, and maybe more...

Even if that is possible, changing many of our dependencies to `peerDependencies` implies that developers must install many more packages by hand. It makes react-admin more intimidating to newcomers. 

On the other hand, the duplicate dependency problem only occurs when two packages depend on non-compatible versions. And even then, it's possible to overcome the problem by using [`resolutions`](https://yarnpkg.com/lang/en/docs/selective-version-resolutions/) in `package.json` to force a single version.

To summarize, I think it's a rare problem that is cumbersome to solve on our side, and that is easy to solve on the developer's side.

So I think we shouldn't do it. But I'm stil open to discussion ;)

**Comment by ljharb** [comment]
@fzaninotto you can make them both peers *and* runtime deps, and then the only time you're forced to do manual work is when you have a peer conflict; as indicated [here](https://github.com/marmelab/react-admin/issues/1683#issuecomment-379805614)

**Comment by fzaninotto** [comment]
@ljharb I couldn't find any documentation on npm or yarn confirming this behavior. Would you have pointers?

**Comment by fzaninotto** [comment]
also, if the change is to duplicate some of our `dependencies` in `peerDependencies`, then it's not a breaking change anymore, and we don't have to deal with it right now.

**Comment by ljharb** [comment]
Adding a peer dependency is always a breaking change.

I don’t know what yarn does; this behavior isn’t documented on npm’s site because overall it’s documentation is poor, but everyone’s is poor around peer deps.

**Comment by fzaninotto** [comment]
After thinking about this a lot, I believe it's a corner case that can be handled using `resolutions`, and fixing it would degrade the newcomer experience too much.

So we won't address it.

**Comment by ljharb** [comment]
only for yarn users, which remain the minority.

**Comment by kopax** [comment]
> After thinking about this a lot, I believe it's a corner case that can be handled using resolutions, and fixing it would degrade the newcomer experience too much.

We are building a stack of modules that depend on ra-core, not using `peerDependencies` is causing us a lot of trouble.

> it can be handled using `resolutions`

This force us to have a lot of fix everywhere.

> it would degrade the newcomer experience too much.

I don't see why, just add the `npm install` command for all the peer, it will just degrade the installation command line by adding more next to `react-admin`. 

Documentation is a nice place to share those kinds of information.

>  So we won't address it.

Please address it :(

**Comment by fzaninotto** [comment]
> I don't see why [it would degrade the newcomer experience]

```diff
-npm install react-admin
+npm install react-admin react-redux react-router material-ui redux-form react-transition-group
```

If I was a newcomer and I saw this installation instructions, I'd run away.

> Every other users who want to build on top of ra-core module will have the issue.

The solution may be to put dependencies as peer in `ra-core`, but leave them as dependencies in `react-admin`. Would that fit your need?

**Comment by ljharb** [comment]
@fzaninotto why would a newcomer care about the length of a copy-pasted command?

Either way, you'd recommend what eslint-config-airbnb does - `npx install-peerdeps react-admin`.

**Comment by fzaninotto** [comment]
> why would a newcomer care about the length of a copy-pasted command

I think a newcomer would care about poor developer experience, poor documentation, and generally losing their time doing things they don't understand. I mean, I would. 

> npx install-peerdeps react-admin

I have never heard about that.

**Comment by ljharb** [comment]
I agree a newcomer cares about these things. But if a step is currently "run this exact command", and it changes to "run this exact command", it has zero negative impact. It's just a different copy and paste.

**Comment by kopax** [comment]
> > I don't see why [it would degrade the newcomer experience]
> 
> ```diff
> -npm install react-admin
> +npm install react-admin react-redux react-router material-ui redux-form react-transition-group
> ```

You don't have to put `react-transition-group` in `peerDependencies`, a `peerDependency` is a singleton dependency, which means they work if they exist only once in the whole.

This is true for packages such as `react`, `styled-components`, `redux` and every dependencies that export a `reducer` (because store must be singleton).

So you should first check if `react-router`, `material-ui`, `redux-form` must have a single copy first before adding them to `peerDependencies`

> > Every other users who want to build on top of ra-core module will have the issue.
> 
> The solution may be to put dependencies as peer in `ra-core`, but leave them as dependencies in `react-admin`. Would that fit your need?

It is a great idea to create a module, then use this module in another module, but that doesn't fix the `peerDependencies` issue for users that still use `react-admin` in another application.

For us it will work and that would make much sens not to have them in `dependencies` in `ra-core`, I personnaly encourage this idea.

But it is still wrong because this would also means that `react-admin` is an application that can be installed in others application. If they are using any of your `peerDependencies` such as `redux` in their application, this will cause errors to them. They will have the choice of, removing `redux` from their `package.json` dependencies, and still import them in their source code. 

Now their linter will cause issues because you should never import dependencies within your source if it is not listed in the `package.json `

If tomorrow you remove `redux` from `react-admin`, you will break their application.

Another way around would consist of exporting every function that your end users want to use from all your `peerDependencies` in your application, so they won't have wrong import, I would personally run away, this is why you should consider setting the `peerDependencies` correctly.

Another option would be to create a module named `react-admin-application`, and then do that. 

You need to decide, is `react-admin` a full framework that exports everything out of the box or is it a module?

**Comment by fzaninotto** [comment]
Every dependency that uses a React context must be a singleton. There is no way around it. So the install command would be long if all of those were a peerDependency. 

Asking people to use anything other than `yarn install` isn't something we're willing to do at this stage - we want to stick with the standards.

We designed react-admin primarily as a framework designed to build complete applications. The fact that it can be used within another app is a bonus, and a secondary consideration.

So again, as the duplicate dependency problem can be solved in userland with `resolution`, and as it occurs in a minority of cases, we won't address it in v3.

**Comment by kopax** [comment]
> So again, as the duplicate dependency problem can be solved in userland with resolution, and as it occurs in a minority of cases, we won't address it in v3.

Could you do it in `ra-core`?

**Comment by ljharb** [comment]
That's not accurate; everything that uses React needs to *have React as a peer dep*, so that *React* is a singleton - it absolutely need not itself be a singleton.

**Comment by kopax** [comment]
@ljharb I think what he meant is that if the context is created and stored within a module that depend of react context, then the module itself needs to be a singleton. But I am not sure if it's true, since I try not to use the context. But for instance, `styled-components` use the context to broadcast the them and perhaps this is why it must be a singleton.

What I am sure of is that: if someone import `redux` to create a `reducer` within a distributed module, and export a reducer in it's api, then the object which store the data is created in a file within this module and thus need to be a peer dependency.

**Comment by ljharb** [comment]
Certainly anything stateful needs to be a singleton and thus be a peer dep of anything that uses it, but that's entirely unrelated to whether it uses React or not.

**Comment by fzaninotto** [comment]
OK, we'll move `ra-core` singleton dependencies to `peerDependencies` + `devDependencies`, and add them as `dependencies` in `react-admin`.

**Comment by fzaninotto** [comment]
@kopax Do you fancy opening a PR for that?

**Comment by kopax** [comment]
Thanks, and of course yes, I will take care of that during the weekend if that's ok. If it must be done earlier, I can arrange myself to do it tonight.

**Comment by kopax** [comment]
Let's continue in https://github.com/marmelab/react-admin/pull/3763

**Comment by kopax** [comment]
This issue should be listed in https://github.com/marmelab/react-admin/issues/3194

**Comment by kopax** [comment]
@fzaninotto I have an idea to make the `Custom App Example` more relevant, it's just an idea to continue the discussion on how to use `react-admin` in another app.

We can duplicate `react-admin` package into `react-admin-custom` using the same source [here](https://github.com/marmelab/react-admin/tree/master/packages/react-admin/src), the difference with `react-admin` and `react-admin-custom` will be that the `devDependencies` and `peerDependencies` will be correctly configured.

Then we can edit `react-admin` to just export what's in `react-admin-custom`. We will set all the `peerDependencies/devDependencies` of `react-admin-custom` in `dependencies` (as in #3194) of `react-admin`. 

The `react-admin` package will have no breaking change and the custom app example will be as simple as using `react-admin`.

This way, the custom app users would have to install the new package that exports everything, and that should work without any breakchange or dependencies collision. Current user and newcomer will keep the same installation line.

**Comment by fzaninotto** [comment]
@kopax Sorry, I didn't understand your point. Can you be more specific?

**Comment by kopax** [comment]
Ok, I'll try again.

The current section `Custom App` is obsolete, because we can't install `react-admin` in a custom app.

Either:

1. Tell people within `Custom App Example` to install `ra-core`, `ra-ui-materialui`, etc. manually

 OR.

1. Create a parent package for `react-admin` and allow `peerDependencies` for most singleton package in it.
1. Move source from `react-admin` into the newly created package, remove `src/*` and export what's in the new package. Fix the `peerDependencies` in `react-admin`. 

This will make things explicit, `react-admin` work in standalone while `react-admin-custom` work for custom app.

For end-users, the one difference will be that custom app users will have to install the parent package of `react-admin` (I called it `react-admin-custom`) and they will have to install all the `peerDependencies` their self.

**Comment by fzaninotto** [comment]
> we can't install react-admin in a custom app

I don't understand why

**Comment by kopax** [comment]
If I install `react-admin` in my custom app and I already have in my dependencies `redux`, then I will have two copy of `redux` in my application.

**Comment by fzaninotto** [comment]
only if you have a different version of redux, am I right?

**Comment by ljharb** [comment]
That's right; and if react-admin has it as a peer dep and a dep, then `npm ls` will fail, and the user's dep graph is invalid, and they'll have to fix that.

**Comment by kopax** [comment]
> That's right; and if react-admin has it as a peer dep and a dep, then `npm ls` will fail, and the user's dep graph is invalid, and they'll have to fix that.

Why would they have a `peerDep` in `dep`?

**Comment by kopax** [comment]
> only if you have a different version of redux, am I right?

I guess yes. (it still depends of the `npm` client you are using, and the version, but normally package get deduped.)

But if people install packages that all depend of `react-admin`, I am not confident that it will exist only once. 

Either way, I am trying to find a solution not to get forced to have all the `peerDependencies` in `dependencies`, and have a support for that.

**Comment by ljharb** [comment]
@kopax by putting it in both, people would only have to manage it themselves if they have a conflicting version.

**Comment by fzaninotto** [comment]
So to summarize, the Custom app documentation describes a working solution, unless developers choose to use a different Redux version. Instead of explaining something complicated with alternative packages, I suggest we just document that gotcha. Using an incompatible Redux version will be possible in 3.0, but not supported, and I don't want to document (=support) that usage.

**Comment by fzaninotto** [comment]
Fixed by #3763

**Comment by kopax** [comment]
Thank you! This is an improvement to us, for v3 and our next packages configurations. 
I look forward to try it in beta.1.
